reading_comprehension/utils.py

```diff
@@ -0,0 +1,134 @@
+import collections
+import torch
+from allennlp.nn.activations import Activation
+
+
+class ResidualWithLayerDropout(torch.nn.Module):
+    """
+    A residual connection with the layer dropout technique `Deep Networks with Stochastic
+    Depth <https://arxiv.org/pdf/1603.09382.pdf>`_ .
+
+    This module accepts the input and output of a layer, decides whether this layer should
+    be stochastically dropped, returns either the input or output + input. During testing,
+    it will re-calibrate the outputs of this layer by the expected number of times it
+    participates in training
+    """
+    def __init__(self, undecayed_dropout_prob=0.5):
+        super(ResidualWithLayerDropout, self).__init__()
+        if undecayed_dropout_prob < 0 or undecayed_dropout_prob > 1:
+            raise ValueError("undecayed dropout probability has to be between 0 and 1, "
+                             "but got {}".format(undecayed_dropout_prob))
+        self.undecayed_dropout_prob = undecayed_dropout_prob
+
+    def extra_repr(self):
+        inplace_str = ', inplace' if self.inplace else ''
+        return 'undecayed_dropout_prob={}{}'.format(self.undecayed_dropout_prob, inplace_str)
+
+    def forward(self, layer_input, layer_output, layer_idx=None, total_layers=None):
+        # pylint: disable=arguments-differ
+        """
+        Apply dropout to this layer, for this whole mini-batch.
+
+        dropout_prob = layer_idx / total_layers * undecayed_dropout_prob if layer_idx and
+        total_layers is specified, else it will use the undecayed_dropout_prob directly.
+
+        Parameters
+        ----------
+        layer_input ``torch.FloatTensor`` required
+            The input tensor of this layer
+        layer_output ``torch.FloatTensor`` required
+            The input tensor of this layer, with the same shape as the layer_input
+        layer_idx ``int``
+            The layer index, starting from 1. This is used to calcuate the dropout prob
+            together with the `total_layers` parameter.
+        total_layers ``int``
+            The total number of layers.
+        Returns
+        -------
+        output: ``torch.FloatTensor``
+            A tensor with the same shape as `layer_input` and `layer_output`.
+        """
+        if layer_idx is not None and total_layers is not None:
+            dropout_prob = 1.0 * self.undecayed_dropout_prob * layer_idx / total_layers
+        else:
+            dropout_prob = 1.0 * self.undecayed_dropout_prob
+        if self.training:
+            if torch.rand(1) < dropout_prob:
+                return layer_input
```

Comment:
Yes, I guess this works probably because it's applied to the residual network. The next layer makes prediction based on all previous layers, so the dropping of one layer might not make much difference.